kv: avoid O(stores) work in updatePausedFollowersLocked

nvb opened the pull request:
Fixes #125308.

This commit avoids O(stores) work in Replica.updatePausedFollowersLocked, which is called on every tick.

Release note: None

nvb committed:
kv: avoid O(stores) work in updatePausedFollowersLocked

Fixes #125308.

This commit avoids O(stores) work in Replica.updatePausedFollowersLocked,
which is called on every tick.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125310)
<!-- Reviewable:end -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125310#-:-NznDCz62YAGfguNSWF1:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

pav-kv commented on pkg/kv/kvserver/replica_raft_overload.go:
```diff
@@ -299,7 +299,7 @@ func (osm *ioThresholds) Replace(
 	r.mu.pausedFollowers = nil
 
-	if ioThresholdMap.NumAbovePauseThreshold() == 0 {
+	if !ioThresholdMap.AnyAbovePauseThreshold(r.descRLocked().Replicas()) {
```
optional: There are a bunch of `Replicas()` calls below. Could be good to put this in a variable, to make it explicit that all the computation here is based on a list of replicas.

nvanbenschoten replied:
Done.

nvb committed:
kv: extract variables in updatePausedFollowersLocked

Refactor.

Epic: None
Release note: None

nvb committed:
roachpb: remove ReplicaSet.AsProto

It is unused and its role is taken over by ReplicaSet.Descriptors().

Epic: None
Release note: None

nvanbenschoten commented:
TFTRs!

bors r+
